dask_expr/repartition.py

```diff
@@ -313,3 +313,45 @@ def _is_single_last_div(x):
                 d[(out2, j - 1)] = (methods.concat, tmp)
             j += 1
         return d
+
+
+def is_broadcastable(dfs, s):
+    """
+    This Series is broadcastable against another dataframe in the sequence
+    """
+
+    def compare(s, df):
+        try:
+            return s.divisions == (df.columns.min(), df.columns.max())
+        except TypeError:
+            return False
+
+    return (
+        s.ndim <= 1
+        and s.npartitions == 1
+        and s.known_divisions
+        and any(compare(s, df) for df in dfs if df.ndim == 2)
+    )
+
+
+def _maybe_align_partitions(args):
+    _is_broadcastable = functools.partial(is_broadcastable, args)
+    dfs = [df for df in args if df.ndim > 0 and not _is_broadcastable(df)]
+    if not dfs:
+        return args
+
+    if all(df.divisions == dfs[0].divisions for df in dfs):
+        return args
+
+    if not all(df.known_divisions for df in dfs):
+        raise ValueError(
+            "Not all divisions are known, can't align "
+            "partitions. Please use `set_index` "
+            "to set the index."
+        )
+
+    divisions = list(unique(merge_sorted(*[df.divisions for df in dfs])))
+    if len(divisions) == 1:  # single value for index
+        divisions = (divisions[0], divisions[0])
+
+    return [Repartition(df, new_divisions=divisions, force=True) for df in dfs]
```

Comment:
This is interesting.  We're taking possibly-lazily-computed-metadata and putting into the expression tree.  This kind of locks us in a little.  

For example, if we, after this, perform a filter and then pass that Filter down through this expression into a ReadParquet then we'll want to change the divisions.  This `Repartition` call will then become inaccurate.

I think that we're discovering a principle that inputs to Expressions should only reflect inputs provided by the user, never something that we compute based on the current form of the expressions.

Reply:
Maybe this makes sense to do if we're deep in the expression process, turning very abstract expressions into more concrete ones.  I don't have confidence yet that we don't want exactly what's here, I mostly want us to agree on the "expression operands should reflect user inputs" principle.

Reply:
I can see your point, I was struggling with something related when starting to look into ``set_index``.

I am just not sure what alternatives we have for operations that require aligning the data. Disallowing them in general could be an option, but that wouldn't be a good solution from a usability perspective. Requiring explicit divisions could be an option for some operations, but this is also not great usability-wise.

> Maybe this makes sense to do if we're deep in the expression process, turning very abstract expressions into more concrete ones. 

I think this touches an interesting point. We could update the ``Repartition`` call when we push something like a filter down. This would keep the divisions updated with the lazily computed ones. Not totally sure whether this would cause problems in other areas though.

Reply:
> We could update the Repartition call when we push something like a filter down.  

That seems hard to track well. 

Instead, my thinking is that we try hard never to create `Repartition(divisions=)` expressions, except after most optimization is done.  I expect these expressions to be the results of operations like the following:

```python
x + y
df.repartition("1 GiB")
```

We should keep these expresssions as close as possible to what the user wrote down:

```python
Add(x, y)
RepartitionSize(df, "1 GiB")
```

It should be easy to optimize around these.  We should try hard not to lower these too early into things like the following:

```python
Add(Repartition(x, divisions=...), Repartition(y, divisions=...))
RepartitionDivisions(df, [...])
```

Once we do that then the expression is kind of locked in and we can't optimize through these terms.  Things are concrete rather than fluid.  

So I think that the `Repartition(divisions=...)` expression is OK, but we want to avoid engaging it as much as possible.  I would be sad if we were lowering to that immediately in user API, for example.  We probably also need to start thinking about when different optimizations happen.  `Repartition(size=...) -> Repartition(divisions=...)` wants to be polite and let other optimizations go first.  Once all of them are done, then it'll go ahead and execute.  We don't currently have any mechanism to control the order of optimizations unfortunately.  

@rjzamora thought about something this a little bit similar (Rick, I'm thinking of simplify vs lower, which I think is a special case of this problem).  Pinging him.